### PR TITLE
[TS migration] Migrate 'getTooltipStyles.js' style to TypeScript

### PR DESCRIPTION
**Pull request by blazejkustra**
### Details

Migration of [src/styles/getTooltipStyles.js](https://github.com/Expensify/App/blob/main/src/styles/getTooltipStyles.js) to Typescript.

### Fixed Issues
$ https://github.com/Expensify/App/issues/24729
PROPOSAL: N/A

### Tests
- [x] Verify that no errors appear in the JS console
- [x] Verify that tooltips are displayed correctly

### Offline tests
N/A

### QA Steps
- [x] Verify that tooltips are displayed correctly (on different screen sizes)
    - Tooltips are used across the app, mostly web and desktop (or only?)
    - Some sample places where tooltips are displayed:
        - Profile 
image
        - User mail
image
        - Chat profile
image
        - Chat Emoji
image
        - Chat send button
image
        - `+` button
image

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/46095609/968c6083-47ad-4a81-acaf-8ab4c6d8281c

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/46095609/6598dc62-310d-4fa1-8a80-c9d292276125

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/46095609/6ee85134-ee1d-4a37-ac0e-25836ec95ea8

</details>

### CODE REVIEW

**Comment by blazejkustra** [comment]
TS migrations are checked by internal engs only, [example](https://github.com/Expensify/App/pull/25744#pullrequestreview-1592099406) (for more info please refer to C+ Slack channel). Could you please approve this PR so we can get an internal engineer to take a look? @burczu

File: src/components/Tooltip/TooltipRenderedOnPageBody.js [hunk 1]
```diff
@@ -86,6 +86,7 @@ const TooltipRenderedOnPageBody = (props) => {
     const {animationStyle, rootWrapperStyle, textStyle, pointerWrapperStyle, pointerStyle} = useMemo(
         () =>
             getTooltipStyles(
+                rootWrapper.current,
```
Comment (fabioh8010) [inline src/components/Tooltip/TooltipRenderedOnPageBody.js:89]: @blazejkustra I saw that you used `RefObject<View>` for the `tooltip` parameter.

```ts
    interface RefObject<T> {
        readonly current: T | null;
    }
```

It is expecting the ref who has `current`, but you are passing only the `current`.

File: src/styles/getTooltipStyles.ts [hunk 2]
```diff
@@ -95,42 +94,49 @@ function isOverlappingAtTop(xOffset, yOffset, tooltip, tooltipTargetWidth, toolt
     return isOverlappingAtTargetCenterX;
 }
 
+type TooltipStyles = {
+    animationStyle: ViewStyle;
+    rootWrapperStyle: Omit<ViewStyle, 'position'> | Pick<CSSProperties, 'position'>;
+    textStyle: TextStyle;
+    pointerWrapperStyle: Omit<ViewStyle, 'position'> | Pick<CSSProperties, 'position'>;
+    pointerStyle: ViewStyle;
```
Comment (fabioh8010) [inline src/styles/getTooltipStyles.ts:102]:
```suggestion
    animationStyle: ViewStyle | CSSProperties;
    rootWrapperStyle: Omit<ViewStyle, 'position'> | Pick<CSSProperties, 'position'>;
    textStyle: TextStyle | CSSProperties;
    pointerWrapperStyle: Omit<ViewStyle, 'position'> | Pick<CSSProperties, 'position'>;
    pointerStyle: ViewStyle | CSSProperties;
```

I think we should allow `CSSProperties` in other styles, shouldn't we?

File: src/components/Tooltip/TooltipRenderedOnPageBody.js (hunk 1 above)
Comment (blazejkustra) [reply src/components/Tooltip/TooltipRenderedOnPageBody.js]: Actually tooltip isn't used at all, so I'll remove the argument.

**Review by fabioh8010 — APPROVED** [review]

**Review by burczu — APPROVED** [review]

**Comment by dangrous** [comment]
Hi! This is the first one of these I've looked at so it might take some time; we're on the merge freeze anyway so it doesn't seem like there's a huge rush.

I will probably ask a number of basic questions, so apologies in advance!

**Comment by fabioh8010** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/20051562/df834de1-1005-4b96-a028-870bbcd257d9

https://github.com/Expensify/App/assets/20051562/c67488ef-9a5f-48ae-882f-fb3f3854c56f

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/20051562/b56097b6-efd2-48e8-8c6e-d4c6635a4c12

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/20051562/504ea1dc-5bc2-4d49-8283-6b3902ef5e81

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/20051562/43ee5070-e3d7-44cf-b90d-763a965cb99c

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/20051562/ffaf20b4-df52-43c4-9fb4-08310c0216c3

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/20051562/47ba2de5-f96a-42eb-9026-3942545c621e

</details>

**Comment by dangrous** [comment]
okay first question - most of this makes sense, but in a couple of the functions we moved around the order of the `tooltip` param compared to where it was before. Was there a reason for this?

**Comment by blazejkustra** [comment]
@dangrous I moved `tooltip` parameter because of an ESLint [rule](https://typescript-eslint.io/rules/default-param-last/). With `tooltip` as last parameter it errors:
image
 
I checked the usage of `getTooltipStyles` function and made sure the usage is changed everywhere.

**Comment by blazejkustra** [comment]
Also just to be double careful, before merging let's make sure this branch is in sync with the newest main @dangrous (we should check that parameters are in the correct order in all function usages)

**Comment by dangrous** [comment]
ah okay cool, thank you for explaining! That makes sense

File: src/styles/getTooltipStyles.ts [hunk 3]
```diff
@@ -61,16 +62,15 @@ function computeHorizontalShift(windowWidth, xOffset, componentWidth, tooltipWid
...
+ * @param xOffset - The distance between the left edge of the window
  *                           and the left edge of the wrapped component.
- * @param {Number} yOffset - The distance between the top edge of the window
+ * @param yOffset - The distance between the top edge of the window
  *                           and the top edge of the wrapped component.
- * @param {Element} [tooltip] - The reference to the tooltip's root element
- * @param {Number} tooltipTargetWidth - The width of the tooltip's target
- * @param {Number} tooltipTargetHeight - The height of the tooltip's target
- * @returns {Boolean}
+ * @param [tooltip] - The reference to the tooltip's root element
```
Comment (dangrous) [inline src/styles/getTooltipStyles.ts:69]: move this one to the top of the param list

**Review by dangrous — CHANGES_REQUESTED** [review]
Just one minor change, otherwise this looks good assuming we double check that we've adjusted all usages of these functions to be in the right order!

**Review by dangrous — APPROVED** [review]
This works for me! Just one final question - do we have an automated TS checker set up yet, or are we relying on us to make sure we catch all the places we need types?

**Comment by blazejkustra** [comment]
@dangrous Yes we have [TS](https://github.com/Expensify/App/actions/runs/6107094381/job/16573425380?pr=26062) and [Lint](https://github.com/Expensify/App/actions/runs/6107094380/job/16573425653?pr=26062) checks

**Comment by dangrous** [comment]
Sweet!

Two more things now:
* Can you merge main into this and retest when you have a moment, as per https://expensify.slack.com/archives/C01GTK53T8Q/p1694405959277169
* Can you expand a bit on the QA steps with some of the places the team can test tooltips? It'll make it easier for the checklist.

**Comment by blazejkustra** [comment]
@dangrous Done!

**Pull request merged**